server/api/src/rsDataObjRename.cpp

```diff
@@ -258,142 +264,174 @@ namespace {
-                 &rodsServerHost );
-    if ( status < 0 ) {
-        return status;
-    }
-    if ( rodsServerHost->localFlag == LOCAL_HOST ) {
-        std::string svc_role;
-        irods::error ret = get_catalog_service_role(svc_role);
-        if(!ret.ok()) {
-            irods::log(PASS(ret));
-            return ret.code();
+        status = getAndConnRcatHost(
+                     rsComm,
+                     MASTER_RCAT,
+                     ( const char* )dataObjRenameInp->srcDataObjInp.objPath,
+                     &rodsServerHost );
+        if ( status < 0 ) {
+            return status;
         }
+        if ( rodsServerHost->localFlag == LOCAL_HOST ) {
+            std::string svc_role;
+            irods::error ret = get_catalog_service_role(svc_role);
+            if(!ret.ok()) {
+                irods::log(PASS(ret));
+                return ret.code();
+            }
 
-        if( irods::CFG_SERVICE_ROLE_PROVIDER == svc_role ) {
-            status = _rsDataObjRename( rsComm, dataObjRenameInp );
-        } else if( irods::CFG_SERVICE_ROLE_CONSUMER == svc_role ) {
-            status = SYS_NO_RCAT_SERVER_ERR;
-        } else {
-            rodsLog(
-                LOG_ERROR,
-                "role not supported [%s]",
-                svc_role.c_str() );
-            status = SYS_SERVICE_ROLE_NOT_SUPPORTED;
+            if( irods::CFG_SERVICE_ROLE_PROVIDER == svc_role ) {
+                status = _rsDataObjRename( rsComm, dataObjRenameInp );
+            } else if( irods::CFG_SERVICE_ROLE_CONSUMER == svc_role ) {
+                status = SYS_NO_RCAT_SERVER_ERR;
+            } else {
+                rodsLog(
+                    LOG_ERROR,
+                    "role not supported [%s]",
+                    svc_role.c_str() );
+                status = SYS_SERVICE_ROLE_NOT_SUPPORTED;
+            }
+        }
+        else {
+            status = rcDataObjRename( rodsServerHost->conn, dataObjRenameInp );
         }
+
+        return status;
     }
-    else {
-        status = rcDataObjRename( rodsServerHost->conn, dataObjRenameInp );
+} // anonymous namespace
+
+int rsDataObjRename(rsComm_t *rsComm, dataObjCopyInp_t *dataObjRenameInp)
+{
+    const auto ec = rsDataObjRename_impl(rsComm, dataObjRenameInp);
+
+    // Update the mtime of the parent collections if a data object was renamed or moved
+    // successfully.
+    if (ec == 0) {
+        namespace fs = irods::experimental::filesystem;
+
+        const auto src_parent_path = fs::path{dataObjRenameInp->srcDataObjInp.objPath}.parent_path();
+        const auto dst_parent_path = fs::path{dataObjRenameInp->destDataObjInp.objPath}.parent_path();
+
+        using std::chrono::system_clock;
+        using std::chrono::time_point_cast;
+
+        const auto mtime = time_point_cast<fs::object_time_type::duration>(system_clock::now());
+
+        try {
+            fs::server::last_write_time(*rsComm, src_parent_path, mtime);
```

Comment:
Very minor thought: Would it be worth (perhaps in the future) having a convenience/helper method to set the last_write_time to now?

Also, the commit message should probably read: "Parent collection's mtime now updated when contents are modified." or something to that effect.

Reply:
that idea has crossed my mind and i'm not sure.  maybe a different function that emphasizes what it is doing?  i don't think `last_write_time` would work given that we can't default the last argument.

i'll update the commit message to be clearer.

Reply:
After coming back and reading this again, I do think convenience functions would be helpful. `std::chrono` is not the easiest library to use when just starting out.

Where would these functions live?

I can imagine having functions that do the following:
- One that returns the current time in various formats.
- One that converts a string (or other type) to an iRODS time type and vice versa.

This should probably have its own issue.

Thoughts on any of that?

Reply:
a generic irods::time or irods::chrono thinger might be useful - but willing to wait for that.  can be a 'later' thing, no problem.